### PR TITLE
New Legend Widget

### PR DESCRIPTION
**Pull request by juandjara**
# Description

Shortcut: https://app.shortcut.com/cartoteam/write/IkRvYyI6I3V1aWQgIjY1MWFiYTI0LTZlMTAtNDJkYi04YzU5LWQ5ZjFjMzhjNzc3MCI=

Revamped LegendWidgetUI with mew design, advanced styling and logic customization.

This PR has a [companion PR](https://github.com/CartoDB/documentation/pull/525) in [documentation repo](https://github.com/CartoDB/documentation/blob/master/app/content/react/library-reference/widgets.md#legendwidget)

## Type of change

- Refactor

# Acceptance

Please describe how to validate the feature or fix

1. check new design on storybook
2. match with figma
3. use in builder

# Basic checklist

- [ ] Good PR name
- [x] Shortcut link
- [ ] Changelog entry
- [ ] Just one issue per PR
- [x] GitHub labels
- [ ] Proper status & reviewers
- [x] Tests
- [ ] Documentation

### CODE REVIEW

File: packages/react-ui/src/widgets/new-legend/LegendLayerTitle.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * @param {string} props.title
+ * @param {boolean} props.visible
+ * @returns {React.ReactNode}
+ */
+export default function LegendLayerTitle({ title, visible }) {
+  const ref = useRef(null);
+  const [isOverflow, setIsOverflow] = useState(false);
+
+  useLayoutEffect(() => {
+    if (visible && ref.current) {
```
Comment (aaranadev) [inline packages/react-ui/src/widgets/new-legend/LegendLayerTitle.js:15]: is it still necessary visible?

File: packages/react-ui/src/widgets/new-legend/LegendLayerVariable.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  [LEGEND_TYPES.CATEGORY]: LegendCategories,
+  [LEGEND_TYPES.ICON]: LegendIcon,
+  [LEGEND_TYPES.BINS]: LegendRamp,
+  [LEGEND_TYPES.PROPORTION]: LegendProportion,
+  [LEGEND_TYPES.CONTINUOUS_RAMP]: LegendRamp
+};
+
+/**
+ * @param {object} props
+ * @param {import('../legend/LegendWidgetUI').LegendLayerVariableData} props.legend - legend variable data.
```
Comment (aaranadev) [inline packages/react-ui/src/widgets/new-legend/LegendLayerVariable.js:18]: Review the imports

File: packages/react-ui/src/widgets/new-legend/LegendLayerVariable.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+ * @returns {React.ReactNode}
+ */
+function LegendUnknown({ legend }) {
+  if (legend.select) {
+    return null;
+  }
+
+  return (
+    <Typography variant='body2' color='textSecondary' component='p'>
+      Legend type {legend.type} not supported
```
Comment (aaranadev) [inline packages/react-ui/src/widgets/new-legend/LegendLayerVariable.js:28]: i18n

File: packages/react-ui/src/widgets/new-legend/LegendLayerVariable.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+/**
+ * @param {object} props
+ * @param {import('../legend/LegendWidgetUI').LegendLayerVariableData} props.legend - legend variable data.
+ * @returns {React.ReactNode}
+ */
+function LegendUnknown({ legend }) {
+  if (legend.select) {
+    return null;
+  }
```
Comment (aaranadev) [inline packages/react-ui/src/widgets/new-legend/LegendLayerVariable.js:24]: is this for basemap?

File: packages/react-ui/src/widgets/new-legend/LegendOpacityControl.js [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+  return (
+    <>
+      <Tooltip title={intlConfig.formatMessage({ id: 'c4r.widgets.legend.opacity' })}>
+        <IconButton
+          size='small'
+          color={open ? 'primary' : 'default'}
+          onClick={() => toggleOpen(!open)}
+        >
+          <svg
```
Comment (aaranadev) [inline packages/react-ui/src/widgets/new-legend/LegendOpacityControl.js:47]: Can you move the SVG to another file?

File: packages/react-ui/src/widgets/new-legend/LegendWidgetUI.styles.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    display: 'flex',
+    justifyContent: 'space-between',
+    position: 'sticky',
+    zIndex: 2,
+    top: 0,
+    background: (theme) => theme.palette.background.paper
+  },
+  legendItemBody: {
+    px: 2
+    // '& [data-testid="categories-legend"] > .MuiGrid-root': {
```
Comment (aaranadev) [inline packages/react-ui/src/widgets/new-legend/LegendWidgetUI.styles.js:43]: if it's not used, remove it

**Review by aaranadev — CHANGES_REQUESTED** [review]

File: packages/react-ui/src/widgets/new-legend/LegendLayerVariable.js (hunk 4 above)
Comment (juandjara) [reply packages/react-ui/src/widgets/new-legend/LegendLayerVariable.js:29]: yes

File: packages/react-ui/src/widgets/legend/LegendCategories.js [hunk 7]
```diff
@@ -1,9 +1,15 @@
...
 import { getPalette } from '../../utils/palette';
 import PropTypes from 'prop-types';
-import Typography from '../../components/atoms/Typography';
-
+import LegendLayerTitle from '../new-legend/LegendLayerTitle';
+import { styles } from '../new-legend/LegendWidgetUI.styles';
+
+/**
+ * @param {object} props
+ * @param {import('../legend/LegendWidgetUI').LegendLayerVariableBase & import('../legend/LegendWidgetUI').LegendCategories} props.legend - legend variable data.
```
Comment (aaranadev) [inline packages/react-ui/src/widgets/legend/LegendCategories.js:10]: Review the imports in JSDOC

File: packages/react-ui/src/widgets/legend/LegendCategories.js [hunk 8]
```diff
@@ -15,24 +21,24 @@ function LegendCategories({ legend }) {
...
-
   return (
-    <Grid container direction='column' spacing={1} data-testid='categories-legend'>
-      {Rows}
-    </Grid>
+    <Box component='ul' data-testid='categories-legend' sx={styles.legendVariableList}>
+      {labels.map((label, idx) => (
+        <LegendCategoriesRow
+          key={label + idx}
+          isMax={false}
```
Comment (aaranadev) [inline packages/react-ui/src/widgets/legend/LegendCategories.js:29]: if it is always false, how useful is it? (I know that it is legacy)

File: packages/react-ui/src/widgets/legend/LegendIcon.js [hunk 9]
```diff
@@ -1,30 +1,32 @@
...
+import { Box } from '@mui/material';
 import PropTypes from 'prop-types';
-import Typography from '../../components/atoms/Typography';
 import { ICON_SIZE_MEDIUM } from '../../theme/themeConstants';
+import { styles } from '../new-legend/LegendWidgetUI.styles';
+import LegendLayerTitle from '../new-legend/LegendLayerTitle';
 
+/**
+ * @param {object} props
+ * @param {import('../legend/LegendWidgetUI').LegendLayerVariableBase & import('../legend/LegendWidgetUI').LegendIcons} props.legend - legend variable data.
```
Comment (aaranadev) [inline packages/react-ui/src/widgets/legend/LegendIcon.js:10]: Review the imports in JSDOC

File: packages/react-ui/src/widgets/legend/LegendProportion.js [hunk 10]
```diff
@@ -28,55 +30,81 @@ const Circle = styled(Box, {
...
+  flexDirection: 'column',
+  justifyContent: 'space-around',
+  gap: 1,
+  flexGrow: 1,
+  flexShrink: 1
+}));
+
+/**
+ * @param {object} props
+ * @param {import('../legend/LegendWidgetUI').LegendLayerVariableBase & import('../legend/LegendWidgetUI').LegendProportion} props.legend - legend variable data.
```
Comment (aaranadev) [inline packages/react-ui/src/widgets/legend/LegendProportion.js:61]: Review the imports in JSDOC

File: packages/react-ui/src/widgets/legend/LegendRamp.js [hunk 11]
```diff
@@ -1,12 +1,18 @@
...
+import { Box, Tooltip, styled } from '@mui/material';
 import PropTypes from 'prop-types';
 import React from 'react';
 import Typography from '../../components/atoms/Typography';
 import { getPalette } from '../../utils/palette';
 import LegendProportion, { getMinMax } from './LegendProportion';
 
+/**
+ * @param {object} props
+ * @param {import('../legend/LegendWidgetUI').LegendLayerVariableBase & import('../legend/LegendWidgetUI').LegendRamp} props.legend - legend variable data.
```
Comment (aaranadev) [inline packages/react-ui/src/widgets/legend/LegendRamp.js:10]: Review the imports in JSDOC

File: packages/react-ui/src/widgets/new-legend/LegendLayerVariable.js [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+  if (legend.isStrokeColor) {
+    return 'c4r.widgets.legend.subtitles.strokeColor';
+  }
+  return 'c4r.widgets.legend.subtitles.color';
+}
+
+/**
+ * @param {object} props
+ * @param {import('../legend/LegendWidgetUI').LegendLayerData} props.layer - Layer object from redux store.
```
Comment (aaranadev) [inline packages/react-ui/src/widgets/new-legend/LegendLayerVariable.js:59]: Review the imports in JSDOC

File: packages/react-ui/src/widgets/new-legend/LegendWidgetUI.js [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+    <Paper elevation={3} sx={rootSx}>
+      {collapsed ? (
+        <Tooltip title={intlConfig.formatMessage({ id: 'c4r.widgets.legend.open' })}>
+          <IconButton onClick={() => onChangeCollapsed(false)}>
+            <LayerIcon />
+          </IconButton>
+        </Tooltip>
+      ) : isMobile ? null : (
+        legendToggleHeader
+      )}
```
Comment (aaranadev) [inline packages/react-ui/src/widgets/new-legend/LegendWidgetUI.js:95]: We should leave the trigger in mobile

**Review by aaranadev — CHANGES_REQUESTED** [review]

File: packages/react-ui/src/localization/id.js [hunk 14]
```diff
@@ -33,14 +33,33 @@ const locales = {
...
+        zoomNote: 'Note: this layer will display at zoom levels',
+        notSupported: 'is not a known legend type',
+        subtitles: {
+          proportion: 'Radius range by',
+          icon: 'Icon based on',
+          strokeColor: 'Stroke color based on',
+          color: 'Color based on'
+        },
+        max: 'Max',
+        min: 'Min'
```
Comment (aaranadev) [inline packages/react-ui/src/localization/id.js:62]: We need to change English to Indonesian

Comment (juandjara) [reply packages/react-ui/src/localization/id.js:62]: All of it? I for sure don't know Indonesian

**Review by aaranadev — CHANGES_REQUESTED** [review]

File: packages/react-ui/src/localization/id.js (hunk 14 above)
Comment (aaranadev) [reply packages/react-ui/src/localization/id.js:62]: ChatGPT? haha

Comment (borja-munoz) [reply packages/react-ui/src/localization/id.js:62]: @gandeszahara can help here

Comment (gandeszahara) [reply packages/react-ui/src/localization/id.js:62]: hi! I can help for the translation. will check and discuss further with Juan

**Review by aaranadev — APPROVED** [review]
LGTM

**Review by aaranadev — APPROVED** [review]

File: packages/react-ui/storybook/.storybook/preview.js [hunk 15]
```diff
@@ -1,14 +1,17 @@
-import React from 'react';
```
Comment (vmilan) [inline packages/react-ui/storybook/.storybook/preview.js:1]: Why do you need to change this file? is too generic

Please, review other stories, at least Legend Category is not working, and I've seen differences in other parts, like this kind of guides: https://cartodb-fb-storybook-react-dev--pr837-feature-new-lege-858d9c1r.web.app/?path=/docs/molecules-avatar--guide

Screenshot 2024-03-15 at 19 24 52

Screenshot 2024-03-15 at 19 24 58

**Review by vmilan — COMMENTED** [review]
I left several comments, but most of them are the same, minor suggestions in different parts

File: packages/react-ui/storybook/.storybook/preview.js (hunk 15 above)
Comment (juandjara) [reply packages/react-ui/storybook/.storybook/preview.js:1]: I don't understand what are you refering to here. If you want, we can have a small meet to review this and explain the changes

File: packages/react-ui/src/widgets/legend/LegendLayer.js [hunk 16]
```diff
@@ -0,0 +1,243 @@
+import React, { useMemo, useRef, useState } from 'react';
+import PropTypes from 'prop-types';
+import { Box, Collapse, IconButton, Tooltip, Typography } from '@mui/material';
```
[the remark this answers is not in the document]
Comment (juandjara) [reply packages/react-ui/src/widgets/legend/LegendLayer.js]: solved, thanks for the heads up

File: packages/react-ui/src/widgets/legend/LegendLayer.js [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+  return (
+    <Box
+      data-testid='legend-layer'
+      aria-label={title}
+      component='section'
+      sx={{
+        '&:not(:first-of-type)': {
+          borderTop: (theme) => `1px solid ${theme.palette.divider}`
+        }
+      }}
```
[the remark this answers is not in the document]
Comment (juandjara) [reply packages/react-ui/src/widgets/legend/LegendLayer.js]: solved, thanks for the heads up

File: packages/react-ui/src/widgets/legend/LegendLayer.js [hunk 18]
```diff
@@ -0,0 +1,243 @@
...
+              }
+            />
+          ))}
+        </LayerVariablesList>
+        {helperText && (
+          <Typography
+            variant='caption'
+            color='textSecondary'
+            component='div'
+            sx={{ p: 2 }}
```
[the remark this answers is not in the document]
Comment (juandjara) [reply packages/react-ui/src/widgets/legend/LegendLayer.js]: solved, thanks for the heads up

File: packages/react-ui/src/widgets/legend/LegendLayer.js [hunk 19]
```diff
@@ -0,0 +1,243 @@
...
+            aria-label={intlConfig.formatMessage({
+              id: collapsed ? 'c4r.widgets.legend.expand' : 'c4r.widgets.legend.collapse'
+            })}
+            disabled={!visible}
+            onClick={() => onChangeCollapsed({ id, collapsed: !collapsed })}
+          >
+            {collapseIcon}
+          </IconButton>
+        )}
+        <Box flexGrow={1} sx={{ minWidth: 0, flexShrink: 1 }}>
```
[the remark this answers is not in the document]
Comment (juandjara) [reply packages/react-ui/src/widgets/legend/LegendLayer.js]: solved, thanks for the heads up

File: packages/react-ui/src/widgets/legend/legend-types/LegendIcon.js [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+ * @param {object} props
+ * @param {import('../LegendWidgetUI').LegendLayerVariableBase & import('../LegendWidgetUI').LegendIcons} props.legend - legend variable data.
+ * @returns {React.ReactNode}
+ */
+function LegendIcon({ legend }) {
+  const { labels = [], icons = [] } = legend;
+  return (
+    <LegendVariableList data-testid='icon-legend'>
+      {labels.map((label, idx) => (
+        <Box key={label} component='li' sx={{ display: 'flex', alignItems: 'center' }}>
```
[the remark this answers is not in the document]
Comment (juandjara) [reply packages/react-ui/src/widgets/legend/legend-types/LegendIcon.js]: solved, thanks for the heads up

File: packages/react-ui/src/widgets/legend/legend-types/LegendRamp.js [hunk 21]
```diff
@@ -26,24 +32,24 @@ function LegendRamp({ isContinuous = false, legend }) {
...
+        <Box maxWidth={240}>
           <Typography variant='overline'>
             You need to specify valid numbers for the labels property
           </Typography>
-        </Grid>
+        </Box>
       ) : (
         <>
-          <Grid container item>
+          <Box sx={{ display: 'flex', pb: 1 }}>
```
[the remark this answers is not in the document]
Comment (juandjara) [reply packages/react-ui/src/widgets/legend/legend-types/LegendRamp.js]: solved, thanks for the heads up

File: packages/react-ui/src/widgets/legend/legend-types/LegendRamp.js [hunk 22]
```diff
@@ -52,14 +58,18 @@ function LegendRamp({ isContinuous = false, legend }) {
                 min={minLabel}
               />
             )}
-          </Grid>
-          <Grid container item justifyContent='space-between'>
-            <Typography variant='overline'>{minLabel}</Typography>
-            <Typography variant='overline'>{maxLabel}</Typography>
-          </Grid>
+          </Box>
+          <Box sx={{ display: 'flex', justifyContent: 'space-between' }}>
```
[the remark this answers is not in the document]
Comment (juandjara) [reply packages/react-ui/src/widgets/legend/legend-types/LegendRamp.js]: solved, thanks for the heads up

File: packages/react-ui/src/widgets/legend/legend-types/LegendProportion.js [hunk 23]
```diff
@@ -0,0 +1,158 @@
...
+
+const Circle = styled(Box, {
+  shouldForwardProp: (prop) => prop !== 'index'
+})(({ index = 0, theme }) => {
+  const width = theme.spacing(sizes[index]);
+  const height = theme.spacing(sizes[index]);
+
+  return {
+    border: `solid 1px ${theme.palette.grey[100]}`,
+    backgroundColor: theme.palette.grey[50],
```
[the remark this answers is not in the document]
Comment (juandjara) [reply packages/react-ui/src/widgets/legend/legend-types/LegendProportion.js]: solved, thanks for the heads up

**Review by vmilan — APPROVED** [review]

**Review by padawannn — CHANGES_REQUESTED** [review]
Testing in Builder, undo this commit https://github.com/CartoDB/carto-react/pull/837/commits/e1f290241a30f92ba8df10d90ef4872a2e47b425 before merge

**Review by padawannn — APPROVED** [review]

**Pull request merged**